File: native/ios/CosmicClock/Sources/CosmicClock/CosmicClockEngine.swift
import Foundation

/// Shared astronomical math — ported from `agent/web/lib/cosmic/math.ts`.
public enum CosmicMath {
    public static let synodicMonthDays = 29.530588853
    public static let precessionPeriodYears = 25_772.0
    private static let knownNewMoon = ISO8601DateFormatter().date(from: "2000-01-06T18:14:00Z")!

    public static func julianDay(_ date: Date) -> Double {
        date.timeIntervalSince1970 / 86_400.0 + 2_440_587.5
    }

    public static func normalizeDeg(_ deg: Double) -> Double {
        var d = deg.truncatingRemainder(dividingBy: 360)
        if d < 0 { d += 360 }
        return d
    }

    public static func sunEclipticLongitudeDeg(jd: Double) -> Double {
        let T = (jd - 2_451_545.0) / 36_525.0
        let L0 = normalizeDeg(280.46646 + 36_000.76983 * T)
        let M = (357.52911 + 35_999.05029 * T) * .pi / 180
        let C = (1.914602 - 0.004817 * T) * sin(M)
            + (0.019993 - 0.000101 * T) * sin(2 * M)
            + 0.000289 * sin(3 * M)
        return normalizeDeg(L0 + C)
    }

    public static func lunarPhaseFraction(_ date: Date) -> Double {
        let elapsed = date.timeIntervalSince(knownNewMoon) / 86_400.0
        let mod = elapsed.truncatingRemainder(dividingBy: synodicMonthDays)
        let positive = mod < 0 ? mod + synodicMonthDays : mod
        return positive / synodicMonthDays
    }

    public static func precessionAngleDeg(_ date: Date) -> Double {
        let years = (julianDay(date) - 2_451_545.0) / 365.25
        return normalizeDeg(years / precessionPeriodYears * 360)
    }

    public static func solarDayAngleDeg(now: Date, solarNoon: Date) -> Double {
        let msPerDay = 86_400_000.0
        var offset = now.timeIntervalSince(solarNoon) * 1000
        offset = offset.truncatingRemainder(dividingBy: msPerDay)
        if offset < 0 { offset += msPerDay }
        let fromNoon = offset > msPerDay / 2 ? offset - msPerDay : offset
        return normalizeDeg(fromNoon / msPerDay * 360)
    }
}

public struct SolarDayEvents: Sendable {
    public var solarNoon: Date
    public var sunrise: Date
    public var sunset: Date
}

public struct CycleLayer: Identifiable, Sendable {
    public let id: String
    public let name: String
    public let tier: Int
    public let angleDeg: Double
    public let phase: Double
}

/// Central engine — one `tick` drives all layers (mirrors TypeScript `CosmicClockEngine`).
public final class CosmicClockEngine: @unchecked Sendable {
    public var latitude: Double
    public var longitude: Double
    public var pressureHpa: Double?
    public var lux: Double?

    public init(latitude: Double, longitude: Double) {
        self.latitude = latitude
        self.longitude = longitude
    }

    public func tick(now: Date = .now) -> [CycleLayer] {
        let sunLon = CosmicMath.sunEclipticLongitudeDeg(jd: CosmicMath.julianDay(now))
        let lunar = CosmicMath.lunarPhaseFraction(now)
        let precession = CosmicMath.precessionAngleDeg(now)
        let breath = pressureHpa.map { 0.5 + max(-1, min(1, ($0 - 1013.25) / 25)) * 0.45 } ?? 0.5
        let light = lux.map { min(1, max(0, log10($0 + 1) / 4)) } ?? 0.5

        return [
            CycleLayer(id: "barometric-breath", name: "Atmospheric Breath", tier: 1, angleDeg: breath * 360, phase: breath),
            CycleLayer(id: "light-spectrum", name: "Light Spectrum", tier: 1, angleDeg: light * 360, phase: light),
            CycleLayer(id: "lunar-synodic", name: "Lunar Synodic", tier: 3, angleDeg: lunar * 360, phase: lunar),
            CycleLayer(id: "sun-ecliptic", name: "Solar Season", tier: 5, angleDeg: sunLon, phase: sunLon / 360),
            CycleLayer(id: "precession", name: "Great Year", tier: 6, angleDeg: precession, phase: precession / 360),
        ]
    }
}

File: agent/web/lib/cosmic/CosmicClockEngine.ts
import { getCycleSnapshot } from "../cycleSystems";
import { computePhases } from "../phase/engine";
import { jdFromDate } from "../phase/timeResolution";
import { computeSolarDayEvents } from "./astronomy";
import {
  muhurtaPhase,
  normalizeDeg,
  PRECESSION_PERIOD_YEARS,
  solarDayAngleDeg,
  tideCycle,
} from "./math";
import { buildSensorSnapshot, luxToSpectrum } from "./sensors";
import type { CosmicClockInput, CosmicClockState, CycleLayer } from "./types";

const CHINESE_ANIMALS = ["Rat", "Ox", "Tiger", "Rabbit", "Dragon", "Snake", "Horse", "Goat", "Monkey", "Rooster", "Dog", "Pig"];

/**
 * Central clock engine: one baseline timestamp drives every concentric cycle layer.
 * Mirrors the native Cosmic Clock architecture (Swift/Kotlin) in TypeScript for the web app.
 */
export class CosmicClockEngine {
  private input: CosmicClockInput;
  private lux: number | null = null;
  private pressureHpa: number | null = null;

  constructor(input: CosmicClockInput) {
    this.input = input;
  }

  setInput(partial: Partial<CosmicClockInput>): void {
    this.input = { ...this.input, ...partial };
  }

  setLux(lux: number | null): void {
    this.lux = lux;
  }

  setPressureHpa(hpa: number | null): void {
    this.pressureHpa = hpa;
  }

  /** Compute full state matrix for a single instant (ms-precision). */
  tick(now: Date = new Date()): CosmicClockState {
    const { lat, lon } = this.input;
    const dayLocal = new Date(now.getFullYear(), now.getMonth(), now.getDate());
    const solar = computeSolarDayEvents(dayLocal, lat, lon);

    // Ephemeris spine: one pure PHASE call replaces the mean-period arithmetic.
    // Drift vs the old math ran up to ~19h for the Moon and days for Mercury.
    // tick() runs at ~60 Hz — scope to the four cycles the clock actually consumes
    // so we don't compute planet-synodic/retrograde every frame.
    const phases = computePhases(jdFromDate(now), {
      lat,
      lon,
      only: ["lunar-synodic", "solar-day", "tropical-year", "precession"],
    });
    const lunarReading = phases.byId["lunar-synodic"];
    const solarDayReading = phases.byId["solar-day"];
    const tropReading = phases.byId["tropical-year"];
    const precReading = phases.byId["precession"];

    const lunarPhase = lunarReading?.phase ?? 0;
    const lunarAngleDeg = normalizeDeg(lunarPhase * 360);
    const tide = tideCycle(now, lunarPhase);
    const muhurta = muhurtaPhase(now, solar.sunrise);
    const sunLon = tropReading
      ? normalizeDeg(Number(tropReading.meta?.solarLongitudeDeg ?? tropReading.angleDeg))
      : 0;
    const precession = precReading?.angleDeg ?? 0;
    // PHASE convention: phase 0 = local midnight, 0.5 = solar noon. The clock's
    // existing convention puts solar noon at 0°, so shift by half a cycle to keep
    // the ring visually unchanged. Fall back to the old solver if PHASE skipped it.
    const solarDayAngle = solarDayReading
      ? normalizeDeg((solarDayReading.phase - 0.5) * 360)
      : solarDayAngleDeg(now, solar.solarNoon);

    const sensors = buildSensorSnapshot({
      lat,
      lon,
      headingDeg: this.input.headingDeg,
      altitudeM: this.input.altitudeM,
      pressureHpa: this.pressureHpa ?? this.input.pressureHpa ?? null,
      lux: this.lux ?? this.input.lux ?? null,
    });
    const ui = luxToSpectrum(sensors.lux);
    const cycles = getCycleSnapshot(now);
    const cnIdx = CHINESE_ANIMALS.indexOf(cycles.chineseZodiac.animal);

    const layers: CycleLayer[] = [
      {
        id: "barometric-breath",
        name: "Atmospheric Breath",
        tier: 1,
        angleDeg: normalizeDeg(sensors.atmosphericBreath * 360),
        phase: sensors.atmosphericBreath,
        color: "#67e8f9",
        meta: {
          pressureHpa: sensors.pressureHpa,
          deltaHpa: sensors.pressureDeltaHpa,
        },
      },
      {
        id: "light-spectrum",
        name: "Light Spectrum",
        tier: 1,
        angleDeg: normalizeDeg(sensors.lightSpectrum * 360),
        phase: sensors.lightSpectrum,
        color: ui.accent,
        meta: { lux: sensors.lux },
      },
      {
        id: "solar-day",
        name: "Terrestrial Day",
        tier: 2,
        angleDeg: solarDayAngle,
        phase: solarDayAngle / 360,
        color: "#f59e0b",
        accuracy: solarDayReading?.accuracy ?? "astronomical",
        claim: "measurement",
        sources: solarDayReading?.sources ?? ["equation of time + observer longitude"],
        meta: {
          solarNoonMs: solar.solarNoon.getTime(),
          sunriseMs: solar.sunrise.getTime(),
          sunsetMs: solar.sunset.getTime(),
        },
      },
      {
        id: "lunar-synodic",
        name: "Lunar Synodic",
        tier: 3,
        angleDeg: lunarAngleDeg,
        phase: lunarPhase,
        color: "#94a3b8",
        accuracy: lunarReading?.accuracy ?? "astronomical",
        claim: "measurement",
        sources: lunarReading?.sources ?? ["astronomy-engine MoonPhase"],
        meta: {
          illumination: Math.round(Number(lunarReading?.meta?.illuminatedFraction ?? lunarPhase) * 1000) / 10,
        },
      },
      {
        id: "tidal",
        name: "Tidal Cycle",
        tier: 3,
        angleDeg: tide.angleDeg,
        phase: tide.angleDeg / 360,
        color: "#0891b2",
        meta: { label: tide.label },
      },
      {
        id: "muhurta",
        name: "Vedic Muhurta",
        tier: 4,
        angleDeg: muhurta.angleDeg,
        phase: (muhurta.index + (muhurta.angleDeg % (360 / 30)) / (360 / 30)) / 30,
        color: "#a855f7",
        meta: { index: muhurta.index, total: 30 },
      },
      {
        id: "tzolkin",
        name: "Tzolkin",
        tier: 4,
        angleDeg: normalizeDeg(((cycles.tzolkin.kin - 1) / 260) * 360),
        phase: cycles.tzolkin.kin / 260,
        color: "#7c3aed",
        meta: {
          kin: cycles.tzolkin.kin,
          sign: cycles.tzolkin.sign,
          tone: cycles.galactic.tone.name,
          tribe: `${cycles.galactic.tribe.color} ${cycles.galactic.tribe.name}`,
          affirmation: cycles.galactic.affirmation,
        },
      },
      {
        id: "chinese-zodiac",
        name: "Chinese Zodiac Day",
        tier: 4,
        angleDeg: normalizeDeg(((cnIdx >= 0 ? cnIdx : 0) / 12) * 360),
        phase: (cnIdx >= 0 ? cnIdx : 0) / 12,
        color: "#dc2626",
        meta: { animal: cycles.chineseZodiac.animal },
      },
      {
        id: "sun-ecliptic",
        name: "Solar Season",
        tier: 5,
        angleDeg: sunLon,
        phase: sunLon / 360,
        color: "#f97316",
        accuracy: tropReading?.accuracy ?? "astronomical",
        claim: "measurement",
        sources: tropReading?.sources ?? ["astronomy-engine SunPosition"],
        meta: { longitude: sunLon, sign: cycles.westernZodiac.sign },
      },
      {
        id: "precession",
        name: "Great Year",
        tier: 6,
        angleDeg: precession,
        phase: precession / 360,
        color: "#6366f1",
        accuracy: precReading?.accuracy ?? "mean-orbit",
        claim: "measurement",
        sources: precReading?.sources ?? ["IAU 2006 precession rate, linear approximation"],
        meta: { periodYears: PRECESSION_PERIOD_YEARS },
      },
    ];

    return {
      timestampMs: now.getTime(),
      now,
      sensors,
      solar,
      solarDayAngleDeg: solarDayAngle,
      lunarPhaseFraction: lunarPhase,
      lunarAngleDeg,
      tideAngleDeg: tide.angleDeg,
      tideLabel: tide.label,
      muhurtaIndex: muhurta.index,
      muhurtaAngleDeg: muhurta.angleDeg,
      sunTropicalLongitudeDeg: sunLon,
      precessionAngleDeg: precession,
      ui,
      layers,
    };
  }
}

export function createCosmicClockEngine(input: CosmicClockInput): CosmicClockEngine {
  return new CosmicClockEngine(input);
}
